### DISCUSSION TITLE
How should we capture telemetry?

### DISCUSSION THREAD

**Discussion by mefellows** [pact-foundation/roadmap#28] [Q&A]
With the introduction of [Google Analytics 4](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=gtag), our current [telemetry system](https://docs.pact.io/telemetry) is no longer capturing usage metrics and we are missing important [information](https://docs.google.com/document/d/1F5kkfhFu7n74T8VKJ3NWWFhnr1WLndDHcBNirCeutIA/edit) to help us make better product decisions. 

Given there is work required to migrate to GA4, it is a timely opportunity to review our approach and tools used. This Q&A invites the community to contribute ideas toward a modern telemetry system that serves the needs of the community, including respecting the privacy of individuals and companies.

**Comment by JP-Ellis** [comment]
Not a direct answer, but it would be good to see how other open source packages capture telemetry. I'll expand on this post as I find more packages which use telemetry.

## Azure CLI

### Data Collected

I was unable to find exactly _what_ it collects.

From inspecting the [source code](https://github.com/Azure/azure-cli/tree/dev/src/azure-cli-telemetry), they upload usage data to [Azure Application Insight](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview).

## AWS SAM CLI

According to AWS' documentation, the CLI collected is:

> **Usage information** – The generic commands and subcommands that customers run.
>
> **Errors and diagnostic information** – The status and duration of commands that customers run, including exit codes, internal exception names, and failures when connecting to Docker.
>
> **System and environment information** – The Python version, operating system (Windows, Linux, or macOS), environment in which the AWS SAM CLI runs (for example, AWS CodeBuild, an AWS IDE toolkit, or a terminal), and hash values of usage attributes.

The [source code](https://github.com/Azure/azure-cli/tree/dev/src/azure-cli-telemetry/azure/cli/telemetry) shows that they upload telemetry to:

```
https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
```

I do not know what is behind that endpoint.

**Reply by mefellows** [reply]
Thanks! Yes, we looked at SAM and dotnet (_updated_: not Azure) of these when creating our own. The dotnet CLI publishes what they send here: https://learn.microsoft.com/en-us/dotnet/core/tools/telemetry
